pkg/codegen/go/gen_program_expressions.go

```diff
@@ -0,0 +1,333 @@
+}
+
+func (g *generator) GenScopeTraversalExpression(w io.Writer, expr *model.ScopeTraversalExpression) {
+	rootName := expr.RootName
+	// TODO splat
+	// if _, ok := expr.Parts[0].(*model.SplatVariable); ok {
+	// 	rootName = "__item"
+	// }
+
+	g.Fgen(w, rootName)
+
+	var objType *schema.ObjectType
+	if resource, ok := expr.Parts[0].(*hcl2.Resource); ok {
+		if schemaType, ok := hcl2.GetSchemaForType(resource.InputType); ok {
+			objType, _ = schemaType.(*schema.ObjectType)
+		}
+	}
+	g.genRelativeTraversal(w, expr.Traversal.SimpleSplit().Rel, expr.Parts, objType)
+}
+
+// GenSplatExpression generates code for a SplatExpression.
+func (g *generator) GenSplatExpression(w io.Writer, expr *model.SplatExpression) { /*TODO*/ }
+
+// GenTemplateExpression generates code for a TemplateExpression.
+func (g *generator) GenTemplateExpression(w io.Writer, expr *model.TemplateExpression) { /*TODO*/ }
+
+// GenTemplateJoinExpression generates code for a TemplateJoinExpression.
+func (g *generator) GenTemplateJoinExpression(w io.Writer, expr *model.TemplateJoinExpression) { /*TODO*/
+}
+
+func (g *generator) GenTupleConsExpression(w io.Writer, expr *model.TupleConsExpression) {
+	g.genTupleConsExpression(w, expr, expr.Type())
+}
+
+// GenTupleConsExpression generates code for a TupleConsExpression.
+func (g *generator) genTupleConsExpression(w io.Writer, expr *model.TupleConsExpression, destType model.Type) {
+	argType := g.argumentTypeName(expr, destType)
+	g.Fgenf(w, "%s{\n", argType)
+	switch len(expr.Expressions) {
+	case 0:
+		// empty array
+		break
+	default:
+		for _, v := range expr.Expressions {
+			g.Fgenf(w, "%v,\n", v)
+		}
+	}
+	g.Fgenf(w, "}")
+}
+
+// GenUnaryOpExpression generates code for a UnaryOpExpression.
+func (g *generator) GenUnaryOpExpression(w io.Writer, expr *model.UnaryOpExpression) { /*TODO*/ }
+
+// argumentTypeName computes the go type for the given expression and model type.
+func (g *generator) argumentTypeName(expr model.Expression, destType model.Type) string {
+	if schemaType, ok := hcl2.GetSchemaForType(destType.(model.Type)); ok {
+		if arrayType, ok := schemaType.(*schema.ArrayType); ok {
+			token := arrayType.ElementType.(*schema.ObjectType).Token
+			tokenRange := expr.SyntaxNode().Range()
+
+			_, module, member, diags := hcl2.DecomposeToken(token, tokenRange)
+			importPrefix := strings.Split(module, "/")[0]
+			contract.Assert(len(diags) == 0)
+			return fmt.Sprintf("%s.%sArray", importPrefix, member)
+		}
+		if objType, ok := schemaType.(*schema.ObjectType); ok {
+			token := objType.Token
+			tokenRange := expr.SyntaxNode().Range()
+
+			_, module, member, diags := hcl2.DecomposeToken(token, tokenRange)
+			importPrefix := strings.Split(module, "/")[0]
+			contract.Assert(len(diags) == 0)
+			return fmt.Sprintf("%s.%s", importPrefix, member)
+		} else if tupType, ok := schemaType.(*schema.ArrayType); ok {
+			fmt.Println(tupType)
+		}
+	}
+
+	return ""
```

Comment:
Not sure if you want to pepper your code with `TODO`s, but this method will need additions for plain types, etc.